sqlmesh/core/model/common.py

```diff
@@ -42,53 +48,80 @@ def make_python_env(
+            if macro_func_or_var.__class__ is d.MacroFunc:
+                name = macro_func_or_var.this.name.lower()
+                if name not in macros:
+                    continue
+
+                # If this macro has been seen before as a non-metadata macro, prioritize that
+                used_macros[name] = (
+                    macros[name],
+                    (used_macros.get(name) or (None, is_metadata))[1],
+                )
+                if name == c.VAR:
+                    args = macro_func_or_var.this.expressions
+                    if len(args) < 1:
+                        raise_config_error("Macro VAR requires at least one argument", path)
+                    if not args[0].is_string:
+                        raise_config_error(
+                            f"The variable name must be a string literal, '{args[0].sql()}' was given instead",
+                            path,
+                        )
+                    used_variables.add(args[0].this.lower())
+            elif macro_func_or_var.__class__ is d.MacroVar:
+                name = macro_func_or_var.name.lower()
+                if name in macros:
+                    # If this macro has been seen before as a non-metadata macro, prioritize that
+                    used_macros[name] = (
+                        macros[name],
+                        (used_macros.get(name) or (None, is_metadata))[1],
+                    )
+                elif name in variables:
+                    used_variables.add(name)
+            elif (
+                isinstance(macro_func_or_var, (exp.Identifier, d.MacroStrReplace, d.MacroSQL))
+            ) and "@" in macro_func_or_var.name:
+                for _, identifier, braced_identifier, _ in MacroStrTemplate.pattern.findall(
+                    macro_func_or_var.name
+                ):
+                    var_name = braced_identifier or identifier
+                    if var_name in variables:
+                        used_variables.add(var_name)
 
     for macro_ref in jinja_macro_references or set():
         if macro_ref.package is None and macro_ref.name in macros:
-            used_macros[macro_ref.name] = macros[macro_ref.name]
+            used_macros[macro_ref.name] = (macros[macro_ref.name], None)
 
-    for name, used_macro in used_macros.items():
+    for name, (used_macro, is_metadata) in used_macros.items():
         if isinstance(used_macro, Executable):
             python_env[name] = used_macro
         elif not hasattr(used_macro, c.SQLMESH_BUILTIN) and name not in python_env:
+            used_macro_func = used_macro.func
+            previous_is_metadata = getattr(used_macro_func, c.SQLMESH_METADATA, None)
+
+            if is_metadata:
+                setattr(used_macro_func, c.SQLMESH_METADATA, is_metadata)
+
             build_env(used_macro.func, env=env, name=name, path=module_path)
+            setattr(used_macro_func, c.SQLMESH_METADATA, previous_is_metadata)
```

Comment:
Why do we set it back? Is it because it will be used by other models and we want to reset the attribute value? This seems like a hack to me. Shouldn't we instead pass the explicit value into `build_env` if needed and only fallback on `getattr(used_macro_func, c.SQLMESH_METADATA, None)` if the explicit value is None?